source/MRViewer/MRSubfeaturePickerWidget.cpp

```diff
@@ -0,0 +1,657 @@
+
+            target->addChild( newSubfeatureData.object );
+
+            ret->subfeatures.push_back( std::move( newSubfeatureData ) );
+        } );
+
+        return ret;
+    };
+
+    std::unordered_set<std::shared_ptr<Object>> visitedObjects;
+
+    auto lambda = [&]( auto lambda, const std::shared_ptr<Object>& cur, ViewportMask visibilityMask, bool isAncillary ) -> void
+    {
+        visibilityMask &= cur->visibilityMask();
+        if ( visibilityMask.empty() )
+            return;
+
+        isAncillary = isAncillary || cur->isAncillary();
+        if ( !allowAncillaryObjects && isAncillary )
+            return;
+
+        if ( auto feature = std::dynamic_pointer_cast<FeatureObject>( cur ) )
+        {
+            if ( !findSubfeature( *feature ) ) // Refuse to recurse into our own subfeatures.
+            {
+                ViewportMask subfeatureVisMask = visibilityMask;
+
+                if ( respectSubfeatureVisualProperty )
+                    subfeatureVisMask &= feature->getVisualizePropertyMask( FeatureVisualizePropertyType::Subfeatures );
+
+                if ( !subfeatureVisMask.empty() )
+                {
+                    if ( !objectPredicate || objectPredicate( *feature, isAncillary ) )
+                    {
+                        auto iter = objects_.find( cur );
+                        bool isNew = iter == objects_.end();
+
+                        bool ok = true;
+                        if ( isNew )
+                        {
+                            if ( auto data = addSubfeatures( feature ) )
+                                iter = objects_.insert_or_assign( cur, std::move( data ) ).first;
+                            else
+                                ok = false;
+                        }
+
+                        if ( ok )
+                            visitedObjects.insert( feature );
+                    }
+                }
+            }
+        }
+
+        for ( const auto& child : cur->children() )
+            lambda( lambda, child, visibilityMask, isAncillary );
+    };
+    lambda( lambda, SceneRoot::getSharedPtr(), getViewerInstance().getPresentViewports(), false );
+
+    for ( auto it = objects_.begin(); it != objects_.end(); )
+    {
+        if ( !visitedObjects.contains( it->first ) )
+        {
+            for ( const SubfeatureData& sub : it->second->subfeatures )
+                sub.object->detachFromParent();
+
+            it = objects_.erase( it );
+        }
+        else
+        {
+            it++;
+        }
+    }
+}
+
+void SubfeaturePickerWidgetLow::removeTemporaryObjects()
+{
+    for ( const auto& elem : objects_ )
+    {
+        for ( const SubfeatureData& sub : elem.second->subfeatures )
+            sub.object->detachFromParent();
+    }
+
+    objects_.clear();
+}
+
+ObjAndPick SubfeaturePickerWidgetLow::pickObject( std::function<bool( const Object& )> pred )
+{
+    ObjAndPick ret{};
+
+    if ( ImGui::GetIO().WantCaptureMouse )
+        return {};
```

Comment:
why check it here? if it i called from `mouseMove` `ImGuiMenu` will take event for itself

Reply:
Ok, removed this one.